controllers/crds/cninode_controller.go

```diff
@@ -91,33 +91,67 @@ func (r *CNINodeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
+	nodeFound := true
+	node := &v1.Node{}
+	if err := r.Client.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			nodeFound = false
+		} else {
+			r.Log.Error(err, "failed to get the node object in CNINode reconciliation, will retry")
+			// Requeue request so it can be retried
+			return ctrl.Result{}, err
+		}
+	}
+
 	if cniNode.GetDeletionTimestamp().IsZero() {
+		shouldPatch := false
+		cniNodeCopy := cniNode.DeepCopy()
 		// Add cluster name tag if it does not exist
 		val, ok := cniNode.Spec.Tags[config.CNINodeClusterNameKey]
 		if !ok || val != r.ClusterName {
-			cniNodeCopy := cniNode.DeepCopy()
 			if len(cniNodeCopy.Spec.Tags) != 0 {
 				cniNodeCopy.Spec.Tags[config.CNINodeClusterNameKey] = r.ClusterName
 			} else {
 				cniNodeCopy.Spec.Tags = map[string]string{
 					config.CNINodeClusterNameKey: r.ClusterName,
 				}
 			}
+			shouldPatch = true
+		}
+		// if node exists, get & add OS label if it does not exist on CNINode
+		if nodeFound {
+			nodeLabelOS := node.ObjectMeta.Labels[config.NodeLabelOS]
```

Comment:
If we have access to node and node metadata has OS info, why can't we use them directly when checking condition for finalizer cleanup?

Reply:
Synced with @sushrk offline, this is for deleted nodes in which case we don't have access the node object. This makes sense to me.